File: problems/链表/160.相交链表.cpp
#include <iostream>
#include <vector>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
};

ListNode* listToLinkedList(vector<int> lst) {
    ListNode* dummy = new ListNode(0);
    ListNode* cur = dummy;
    for (int val : lst) {
        cur->next = new ListNode(val);
        cur = cur->next;
    }
    return dummy->next;
}

vector<int> linkedListToList(ListNode* node) {
    vector<int> result;
    while (node) {
        result.push_back(node->val);
        node = node->next;
    }
    return result;
}

ListNode* getIntersectionNode(ListNode* headA, ListNode* headB) {
    if (!headA || !headB) return nullptr;
    ListNode* pa = headA;
    ListNode* pb = headB;
    while (pa != pb) {
        pa = pa ? pa->next : headB;
        pb = pb ? pb->next : headA;
    }
    return pa;
}

void printList(ListNode* node) {
    cout << "[";
    while (node) {
        cout << node->val;
        if (node->next) cout << ",";
        node = node->next;
    }
    cout << "]";
}

int main() {
    // 测试用例1: 相交于节点 8
    ListNode* common = new ListNode(8);
    common->next = new ListNode(4);
    common->next->next = new ListNode(5);
    ListNode* headA = new ListNode(4);
    headA->next = new ListNode(1);
    headA->next->next = common;
    ListNode* headB = new ListNode(5);
    headB->next = new ListNode(0);
    headB->next->next = new ListNode(1);
    headB->next->next->next = common;
    ListNode* result = getIntersectionNode(headA, headB);
    cout << "测试1 - 相交链表:" << endl;
    cout << "  输出: " << (result ? to_string(result->val) : "null") << endl;

    // 测试用例2: 不相交
    ListNode* headC = listToLinkedList({1, 2, 3});
    ListNode* headD = listToLinkedList({4, 5, 6});
    ListNode* result2 = getIntersectionNode(headC, headD);
    cout << "\n测试2 - 不相交链表:" << endl;
    cout << "  输出: " << (result2 ? to_string(result2->val) : "null") << endl;

    // 测试用例3: 完全重合
    ListNode* headE = listToLinkedList({1, 2, 3});
    ListNode* result3 = getIntersectionNode(headE, headE);
    cout << "\n测试3 - 完全重合链表:" << endl;
    cout << "  输出: " << (result3 ? to_string(result3->val) : "null") << endl;

    return 0;
}

File: problems/链表/160.相交链表.py
#!/usr/bin/env python3
"""
160. 相交链表 (Intersection of Two Linked Lists)
题目: 给你两个单链表的头节点 headA 和 headB，请你返回相交链表的头节点。
如果两个链表没有交点，返回 null。
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def get_intersection_node(headA: ListNode, headB: ListNode) -> ListNode:
    if not headA or not headB:
        return None
    pa, pb = headA, headB
    while pa != pb:
        pa = pa.next if pa else headB
        pb = pb.next if pb else headA
    return pa

def list_to_linkedlist(lst):
    dummy = ListNode(0)
    cur = dummy
    for val in lst:
        cur.next = ListNode(val)
        cur = cur.next
    return dummy.next

def linkedlist_to_list(node):
    result = []
    while node:
        result.append(node.val)
        node = node.next
    return result

if __name__ == "__main__":
    # 测试用例1: 相交于节点 8
    # A: 4->1->8->4->5, B: 5->0->1->8->4->5 (相交于 8)
    common = ListNode(8, ListNode(4, ListNode(5)))
    headA = ListNode(4, ListNode(1, common))
    headB = ListNode(5, ListNode(0, ListNode(1, common)))
    result = get_intersection_node(headA, headB)
    print("测试1 - 相交链表:")
    print(f"  输出: {result.val if result else None}")  # 8

    # 测试用例2: 不相交
    headC = list_to_linkedlist([1, 2, 3])
    headD = list_to_linkedlist([4, 5, 6])
    result2 = get_intersection_node(headC, headD)
    print("\n测试2 - 不相交链表:")
    print(f"  输出: {result2}")  # None

    # 测试用例3: 完全重合
    headE = list_to_linkedlist([1, 2, 3])
    result3 = get_intersection_node(headE, headE)
    print("\n测试3 - 完全重合链表:")
    print(f"  输出: {result3.val if result3 else None}")  # 1
